x-pack/plugins/rule_registry/README.md

```diff
@@ -0,0 +1,68 @@
+    type: 'date',
+    array: false,
+    required: true,
+  }
+}
+```
+
+ECS mappings are generated via a script in the rule registry plugin directory. These mappings are available in x-pack/plugins/rule_registry/server/generated/ecs_field_map.ts.
+
+To pick many fields, you can use `pickWithPatterns`, which supports wildcards with full type support.
+
+If a registry is created, it will initialise as soon as the core services needed become available. It will create a (versioned) template, alias, and ILM policy, but only if these do not exist yet.
+
+### Rule registry client
+
+The rule registry client can either be injected in the executor, or created in the scope of a request. It exposes a `search` method and a `bulkIndex` method. When `search` is called, it first gets all the rules the current user has access to, and adds these ids to the search request that it executes. This means that the user can only see data from rules they have access to.
+
+Both `search` and `bulkIndex` are fully typed, in the sense that they reflect the mappings defined for the registry.
+
+### Schema
+
+The following fields are available in the root rule registry:
+
+- `@timestamp`: the ISO timestamp of the alert event. For the lifecycle rule type helper, it is always the value of `startedAt` that is injected by the Kibana alerting framework.
+- `event.kind`: signal (for the changeable alert document), state (for the state changes of the alert, e.g. when it opens, recovers, or changes in severity), or metric (individual evaluations that might be related to an alert).
+- `event.action`: the reason for the event. This might be `open`, `close`, `active`, or `evaluate`.
+- `tags`: tags attached to the alert. Right now they are copied over from the rule.
+- `rule.id`: the identifier of the rule type, e.g. `apm.transaction_duration`
+- `rule.uuid`: the saved objects id of the rule.
+- `rule.name`: the name of the rule (as specified by the user).
+- `rule.category`: the name of the rule type (as defined by the rule type producer)
+- `kibana.rac.producer`: the producer of the rule type. Usually a Kibana plugin. e.g., `APM`.
+- `kibana.rac.alert.id`: the id of the alert, that is unique within the context of the rule execution it was created in. E.g., for a rule that monitors latency for all services in all environments, this might be `opbeans-java:production`.
+- `kibana.rac.alert.uuid`: the unique identifier for the alert during its lifespan. If an alert recovers (or closes), this identifier is re-generated when it is opened again.
+- `kibana.rac.alert.status`: the status of the alert. Can be `open` or `closed`.
+- `kibana.rac.alert.start`: the ISO timestamp of the time at which the alert started.
+- `kibana.rac.alert.end`: the ISO timestamp of the time at which the alert recovered. 
+- `kibana.rac.alert.duration.us`: the duration of the alert, in microseconds. This is always the difference between either the current time, or the time when the alert recovered. 
```

Comment:
Is `kibana.rac.alert.duration.us` necessary? Isn't it simple enough to calculate the diff between `kibana.rac.alert.start` and `kibana.rac.alert.end`?

> This is always the difference between either the current time

How does this work in practice? Will `duration.us` be updated regularly if no `kibana.rac.alert.end` is given?

Reply:
> Is kibana.rac.alert.duration.us necessary? Isn't it simple enough to calculate the diff between kibana.rac.alert.start and kibana.rac.alert.end?

It's complicated at query time. We discussed this briefly in the terminology meeting last week, and consensus was that precalculating this is preferred.

> How does this work in practice? Will duration.us be updated regularly if kibana.rac.alert.end is empty?

Yes, at each every execution.

Reply:
w/r/t `duration` - For the original event log work, I had decided to not use duration, since we were providing start and end, but the advice from ECS was to add duration anyway, so users didn't have to do the math.  Perhaps a runtime field here is a possibility?

We also did not multi-field `duration` (eg, `duration.us`), which is nanoseconds in ECS, but in practice, I'm not sure microseconds is practical for humans either.  For that matter, milliseconds may not be useful to humans, but is super-useful for devs for diagnostic purposes.

w/r/t "is duration updated regularly if end is empty" - we'll have a similar issue when we start adding duration to the event log "active-instance" docs - currently these only have the timestamp, but we'd like to add the `start` so we can calculate the "current duration" (or duration in the context of the document).  Do we add an `end` field so `start` / `end` / `duration` make sense, in the context of just looking at those fields?  The alert may still be active, so adding an `end` may be semantically confusing, because it's not really "the end". 

I'm in favor of just adding all these fields, and you need to be aware contextually that `end` for these documents doesn't indicate the alert is no longer triggering, it's just a reflection of the state at the time the document was generated - 'it had been active for X minutes at time Y".

Reply:
@pmuellr agreed re micro vs milli, micro doesn't really make sense for executors. I just copied duration.us from APM mappings 😄 I'll add it to the list-of-things-we-should-probably-change.

> I'm in favor of just adding all these fields, and you need to be aware contextually that end for these documents doesn't indicate the alert is no longer triggering, it's just a reflection of the state at the time the document was generated - 'it had been active for X minutes at time Y".

💯